[sival/aes]  Introduce aes_force_prng_reseed_test and aes_prng_reseed_test

github-gcontributor opened the pull request:
[sival/aes] Add aes_prng_reseed,aes_prng_force_reseed
    
    This commit introduces the AES PRNG reseed test (chip_sw_aes_prng_reseed)
    and (chip_sw_aes_force_prng_reseed) and AES FORCE PRNG reseed, to handle the
    reseed process by disabling entrpy complex and triggering PRNG reseed
    accordingly to test the actual behavior of the AES module's PRNG reseeding in
    ECB mode.
    
    - Updated the test plan descriptions in `chip_aes_testplan.hjson` to match
      the implemented tests. Added missing test feature.
    - Resoloved latest code suggestions provided in PR #24574 and updated
      code with suggestions from other thread (using TRY and remove .c suffix
      in bazel tag).
    - Single commit for a cleaner history.
    - Resolve multiple commits.
    
    See: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson
    Closes Issue #24857
    Closes Issue #24899
    PR #24574

github-gcontributor commented:
@moidx ,

Updated with feedback and as per format/style.
https://github.com/github-gcontributor/opentitan/blob/vt_checkins/sw/device/tests/aes_force_prng_reseed_test.c 
Please review.

moidx commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,264 @@
+  LOG_INFO(
+      "AES transaction configured: ECB mode, 128-bit key, manual operation.");
+
+  // Disable EDN and CSRNG to simulate stalling
+  LOG_INFO("Disabling EDN0, EDN1...");
+  CHECK_DIF_OK(dif_edn_stop(&edn0));
+  CHECK_DIF_OK(dif_edn_stop(&edn1));
+  busy_spin_micros(500);
```
Use the following function from `entropy_testutils`:

```c
TRY(entropy_testutils_stop_all());
```

github-gcontributor replied:
Noted

moidx commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,264 @@
+  LOG_INFO("Conditionally checking AES status after stopping EDN and CSRNG.");
+  LOG_INFO("Waiting for AES to stall after PRNG reseed request...");
+  bool stall = false;
+  CHECK_DIF_OK(dif_aes_get_status(&aes, kDifAesStatusStall, &stall));
```
The stall condition will only occur after you switch the key. You may have to load a couple of different keys because the PRNG may already be reseeded the first time you call it. 

This check should go below after `dif_aes_start(...)`.

github-gcontributor replied:
Noted.  Pushing "waiting for the stall" logic after dif_aes_start()

github-gcontributor replied:
For loading keys more than once. 
Steps will be as below:
generate first key = key
Generate key_share0[i] = kAesModesKey128[i] ^ kKeyShare1[i]
Load first key using dif_aes_start()
Then dif_aes_trigger()

generate second key = key2
Generate key2_share0[i] = kAesModesKey128[i] ^ kKeyShare1[i]+1 --> Different key generation

wait for stall
then enable entropy complex
load data

Load second key using dif_aes_start() later after enabling 
Then dif_aes_trigger()
Please advice.

moidx commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,264 @@
+  // T4: dif_aes_start
+  CHECK_DIF_OK(dif_aes_start(&aes, &transaction, &key, NULL));
+
+  LOG_INFO("Disabling CSRNG...");
+  CHECK_DIF_OK(dif_csrng_stop(&csrng));
+  CHECK_DIF_OK(dif_entropy_src_stop(&entropy_src));
+
+  // Re-enable Entropy Source and CSRNG in the proper sequence
+  LOG_INFO("Re-initializing Entropy Source and CSRNG...");
+
+  // Initialize and configure Entropy Source
+  dif_entropy_src_config_t entropy_src_config =
+      entropy_testutils_config_default();
+  CHECK_DIF_OK(dif_entropy_src_configure(&entropy_src, entropy_src_config,
+                                         kDifToggleEnabled));
+
+  // Initialize and start CSRNG
+  CHECK_DIF_OK(dif_csrng_configure(&csrng));
+  // set_edn_auto_mode();
+  LOG_INFO("Setting EDN to auto mode...");
+  TRY(entropy_testutils_auto_mode_init());
```
All this code can be replaced for a single call:

```c
TRY(entropy_testutils_auto_mode_init());
```

github-gcontributor replied:
Noted

moidx commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,264 @@
+  // CHECK_DIF_OK(entropy_testutils_auto_mode_init());
+  monitor_edn_status(&edn0, &edn1);
+
+  // Trigger PRNG reseed request
+  LOG_INFO("Triggering PRNG reseed...");
+  CHECK_DIF_OK(dif_aes_trigger(&aes, kDifAesTriggerPrngReseed));
```
I would expect to see a stall if entropy complex is disabled and you call this command. It may be a good idea to test this as well.

github-gcontributor replied:
Old comment

moidx commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,264 @@
+  // T8: Load the plain text to trigger the encryption operation
+  CHECK_DIF_OK(dif_aes_load_data(&aes, in_data_plain));
+
+  // Trigger AES reseed request, specific for manual operation case
```
`kDifAesTriggerStart` is not the same as `kDifAesTriggerPrngReseed`.

github-gcontributor replied:
Yes. Can you clarify more on this review comment wrt code?

moidx commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,264 @@
+
+  busy_spin_micros(500);
+
+  LOG_INFO("Disabling CSRNG...");
+  CHECK_DIF_OK(dif_csrng_stop(&csrng));
+  CHECK_DIF_OK(dif_entropy_src_stop(&entropy_src));
+
+  // Re-enable Entropy Source and CSRNG in the proper sequence
+  LOG_INFO("Re-initializing Entropy Source and CSRNG...");
+  // Initialize and configure Entropy Source
+  dif_entropy_src_config_t entropy_src_config_1 =
+      entropy_testutils_config_default();
+  CHECK_DIF_OK(dif_entropy_src_configure(&entropy_src, entropy_src_config_1,
+                                         kDifToggleEnabled));
+  // Initialize and start CSRNG
+  CHECK_DIF_OK(dif_csrng_configure(&csrng));
+
+  LOG_INFO("Setting EDN to auto mode...");
+  TRY(entropy_testutils_auto_mode_init());
+  // CHECK_DIF_OK(entropy_testutils_auto_mode_init());
```
It is not clear to me why we are doing this?

github-gcontributor replied:
Noted.. 
Removing the redundant logic...
We already enabled entropy complex earlier...

moidx commented:
Hi @github-gcontributor, Can you also do the following:

1. Add Bazel target to `sw/device/tests/BUILD`
2. Update the commit title and text to describe the changes being introduced in this PR.
3. Update the PR title and description to follow the commit details. 
4. You will also have to setup your `Name` and `Email` in gitgub to be able to sign your commit. This is the reason one of the CI checks is failing. This is needed to meet the CLA guidelines.

github-gcontributor commented:
@moidx,
Thanks for the review.
Made changes to the code as per the comments.
i.e. The Test checks for AES Stall with Disabled Entropy Complex (and not just edn)
Added BUILD file with target

Please provide feedback on the changes..

github-gcontributor commented:
> Hi @github-gcontributor, Can you also do the following:
> 
> 1. Add Bazel target to `sw/device/tests/BUILD`
> 2. Update the commit title and text to describe the changes being introduced in this PR.
> 3. Update the PR title and description to follow the commit details.
> 4. You will also have to setup your `Name` and `Email` in gitgub to be able to sign your commit. This is the reason one of the CI checks is failing. This is needed to meet the CLA guidelines.

Regarding point 4, 
I tried making the changes as per the instruction.  Using my corporate id, does not allow me to proceed checkin as it would expose the privacy.  As a second step, I added "<number>+github-gcontributor@users.noreply.github.com" as mentioned in my github steps to protect privacy.  However it still asking showing up issue in quick lint check.  I added my personal id.

a-will commented:
> > 4. You will also have to setup your `Name` and `Email` in gitgub to be able to sign your commit. This is the reason one of the CI checks is failing. This is needed to meet the CLA guidelines.
> 
> Regarding point 4, I tried making the changes as per the instruction. Using my corporate id, does not allow me to proceed checkin as it would expose the privacy. As a second step, I added "[+github-gcontributor@users.noreply.github.com](mailto:+github-gcontributor@users.noreply.github.com)" as mentioned in my github steps to protect privacy. However it still asking showing up issue in quick lint check.
> 
> Still issue is persistent: """ remote: error: GH007: Your push would publish a private email address. remote: You can make your email public or disable this protection by visiting: remote: [http://github.com/settings/emails](https://github.com/settings/emails) To github.com:github-gcontributor/opentitan.git ! [remote rejected] vt_checkins -> vt_checkins (push declined due to email privacy restrictions) error: failed to push some refs to 'github.com:github-gcontributor/opentitan.git' """ I added my personal id. With non corporate id, I am disabling the "Keep my email addresses private" option now. Looks like it is allowing me to chekcin now. Waiting for the quick lint

You are contributing to a public repo with a CLA from your company. The corporate email must be exposed in commits, else the contribution cannot be accepted. You must uncheck "Block command line pushes that expose my email" (or else get someone else from your organization to submit the change).

Some people choose to have separate accounts for corporate work, and this enabled them to keep this safeguard for their personal account. I'm not sure if there is a way to make it more granular (like by repo or something).

github-gcontributor commented:
> > > 4. You will also have to setup your `Name` and `Email` in gitgub to be able to sign your commit. This is the reason one of the CI checks is failing. This is needed to meet the CLA guidelines.
> > 
> > 
> > Regarding point 4, I tried making the changes as per the instruction. Using my corporate id, does not allow me to proceed checkin as it would expose the privacy. As a second step, I added "[+github-gcontributor@users.noreply.github.com](mailto:+github-gcontributor@users.noreply.github.com)" as mentioned in my github steps to protect privacy. However it still asking showing up issue in quick lint check.
> > Still issue is persistent: """ remote: error: GH007: Your push would publish a private email address. remote: You can make your email public or disable this protection by visiting: remote: [http://github.com/settings/emails](https://github.com/settings/emails) To github.com:github-gcontributor/opentitan.git ! [remote rejected] vt_checkins -> vt_checkins (push declined due to email privacy restrictions) error: failed to push some refs to 'github.com:github-gcontributor/opentitan.git' """ I added my personal id. With non corporate id, I am disabling the "Keep my email addresses private" option now. Looks like it is allowing me to chekcin now. Waiting for the quick lint
> 
> You are contributing to a public repo with a CLA from your company. The corporate email must be exposed in commits, else the contribution cannot be accepted. You must uncheck "Block command line pushes that expose my email" (or else get someone else from your organization to submit the change).
> 
> Some people choose to have separate accounts for corporate work, and this enabled them to keep this safeguard for their personal account. I'm not sure if there is a way to make it more granular (like by repo or something).

Sure, in that case let me change my id to corporate email

pamaury commented:
Hi @github-gcontributor, could you please edit the title of the pull request to something that reflects its content? This makes it easier for other contributors to know what the PR is about.

github-gcontributor commented:
> Hi @github-gcontributor, could you please edit the title of the pull request to something that reflects its content? This makes it easier for other contributors to know what the PR is about.


Title updated

nasahlpa commented:
Could you please add the description what this commit is doing to the git commit message. Please also update the title of the git commit message. This title should contain a [subject tag](https://opentitan.org/book/doc/contributing/detailed_contribution_guide/index.html#subject-tags).

github-gcontributor commented:
> Could you please add the description what this commit is doing to the git commit message. Please also update the title of the git commit message. This title should contain a [subject tag](https://opentitan.org/book/doc/contributing/detailed_contribution_guide/index.html#subject-tags).

Done

nasahlpa commented:
Thanks! Would you mind looking into the review comments I've added? Especially the question regarding why the test seems to fail should be addressed.

github-gcontributor commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,248 @@
+      if (stall) {
+        LOG_INFO("AES encryption has stalled as expected.");
+      } else {
+        LOG_WARNING(
```
As per the description of the testcase, test run is supposed to generate stall, however the stall bit is not getting asserted.  As per my observation, this is a bug that I would like to file once confirmed.

github-gcontributor replied:
Noted! (see earlier comment reply).

github-gcontributor commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,248 @@
+      if (output_valid) {
+        LOG_INFO("AES encryption has output_valid as expected.");
+      } else {
+        LOG_ERROR(
```
As per my understanding of the module, the normal operation expects output_valid. However, when entropy complex is disabled output valid is not working as expected. Expected behavior is to see stall enabled.  This is an issue.

github-gcontributor replied:
Sure. Noted.  Additionally, with your valid inputs and internal discussion, shall make changes as below as correct operation procedure for the test: 
* Number of blocks set to 68 (4 more to see if tests auto generates reseed post 64Block)
* Remove the trigger prng reseed  (Not required for Auto test)
* Commenting out the enabling of the entropy complex at block 32 -- correct test procedure (Enable once entire 64 block of data is processed)
* Enabling the entropy complex only after entire encryption process completes (after 68 blocks) -- correct procedure.
* Remove code dif_aes_end,  new_key_generation and dif_aes_start - incorrect logic for auto mode
* Remove check for Stall signal bit set (remove Stall check logic from the code) -- as per correct procedure of the test
* After entropy disabled, wait for output valid normally
* Create CL for "Change of test case description-aes_prng_reseed_test" --> Ambiguity in interpreting encryption stalling with STALL register bit

github-gcontributor replied:
Created issue: https://github.com/lowRISC/opentitan/issues/24857

github-gcontributor commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,248 @@
+  LOG_INFO("End AES encryption operation");
+  CHECK_DIF_OK(dif_aes_end(&aes));
+
+  // Wait for AES to be idle before starting decryption
```
Typo,
Will change the comment, "// Wait for AES to be idle before starting decryption" to "// Wait for AES to be idle to check aes encryption process has ended"

github-gcontributor commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  // Start encryption process
+  for (int i = 0; i < kAesNumBlocks; ++i) {
```
The test is an expanded form of the function (dif_aes_process_data) for better understandability of the encryption data flow.  The steps are well simplified in this process.  I have purposely adopted this approach to uncover the stall issue that we see in the subsequent steps.  I would be glad to know if there is any discrepancy in the steps.

nasahlpa replied:
Alright, thanks for the explanation.

github-gcontributor commented on sw/device/tests/BUILD:
```diff
@@ -6279,3 +6279,61 @@ opentitan_test(
     ],
 )
+
+opentitan_test(
```
Noted!

github-gcontributor commented on sw/device/tests/BUILD:
```diff
@@ -6279,3 +6279,61 @@ opentitan_test(
+    ],
+)
+
+opentitan_test(
```
Noted!

github-gcontributor commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,248 @@
+  generate_new_key(&key, 0);
+
+  // Prepare transaction
+  transaction.operation = kDifAesOperationEncrypt;
+  transaction.mode = kDifAesModeEcb;
+  transaction.key_len = kDifAesKey128;
+  transaction.key_provider = kDifAesKeySoftwareProvided;
+  transaction.mask_reseeding = kDifAesReseedPer64Block;
+  transaction.manual_operation = kDifAesManualOperationAuto;
+  transaction.reseed_on_key_change = true;
+  transaction.force_masks = false;
+  transaction.ctrl_aux_lock = false;
```
Noted!

github-gcontributor commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,248 @@
+
+  // Start the AES operation
+  // Wait for AES to be idle before starting encryption
+  aes_idle = false;
+  do {
+    CHECK_DIF_OK(dif_aes_get_status(&aes, kDifAesStatusIdle, &aes_idle));
+  } while (!aes_idle);
```
I want to be consistent in using "DIF" functions instead of "TESTUTILS".  (I shall remove testutils in this case)
I shall accept the code as per your suggestion if its hindering any process flow.

nasahlpa replied:
Thanks for the explanation, fine for me.

engdoreis commented on sw/device/tests/BUILD:
Could you please re-order the commit message (first commit)? The Signed-off line should be after the description.

engdoreis replied:
Please drop the second commit (It's a merge commit). You should use `git rebase` in order to get the latests commit from master.

github-gcontributor replied:
Noted

engdoreis commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,271 @@
+static dif_aes_t aes;
+dif_aes_transaction_t transaction;
+dif_aes_key_share_t key;
+enum {
+  kAesNumBlocks = 8,
+};
+// NumBytes should be 1 i.e. 8 bits
+enum {
+  kDifAesBlockNumBytes = 1,
+};
+enum {
+  disable_entropy_after_block = 0,
+};
+enum {
+  disable_entropy_at_start_en = 1,
+};
+enum {
+  generate_new_key_at_block = 3,
+};
```
```suggestion
enum {
  kAesNumBlocks = 8,
// NumBytes should be 1 i.e. 8 bits
  kDifAesBlockNumBytes = 1,
  kDisableEntropyAfterBlock = 0,
  kDisableEntropyAtStartEn = 1,
  kGenerateNewKeyAtBlock = 3,
};
```

github-gcontributor replied:
Noted!

engdoreis commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,271 @@
+  // Create plaintext with incremental data
+  for (int i = 0; i < kAesNumBlocks; ++i) {
+    for (int j = 0; j < kDifAesBlockNumBytes; ++j) {
+      plain_text[i].data[j] = (uint8_t)((i * kDifAesBlockNumBytes + j) & 0xFF);
```
Maybe use entropy like here:
https://github.com/lowRISC/opentitan/blob/831113c05ab4620c551806e26ba6963e50eec910/sw/device/tests/pmod/i2c_host_fram_test.c#L145-L149

github-gcontributor replied:
Updated code snippet::

#include "sw/device/lib/dif/dif_rv_core_ibex.h"
#include "sw/device/lib/testing/rv_core_ibex_testutils.h"
  
  // Initialize plaintext data dynamically
  // Create plaintext with random data
  dif_rv_core_ibex_t rv_core_ibex;
  CHECK_DIF_OK(dif_rv_core_ibex_init(
      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR),
      &rv_core_ibex));
  for (int i = 0; i < kAesNumBlocks; ++i) {
    for (int j = 0; j < kDifAesBlockNumBytes; ++j) {
      uint32_t rand;
      TRY(rv_core_ibex_testutils_get_rnd_data(&rv_core_ibex, 2000, &rand));
      plain_text[i].data[j] = rand;
    }
  }

github-gcontributor replied:
The snippet is working code.  Please let me know your views on it.

github-gcontributor replied:
Code updated and checked in

engdoreis replied:
I don't see the change.

github-gcontributor replied:
Done

nasahlpa commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,228 @@
+            "disabled at block %d",
+            i);
+      } else {
+        LOG_ERROR("ERROR:AES encryption did not generate output_valid");
```
When running your code, I again get the `ERROR:AES encryption did not generate output_valid` message. Is this intended? When disabling the entropy complex, shouldn't output valid not be 0?

github-gcontributor replied:
Issue: 
https://github.com/lowRISC/opentitan/issues/24899

github-gcontributor replied:
Hi Pascal,
Wondering if are you able to run the test at your end?

nasahlpa replied:
Hi
Sorry, I am currently OoO - will be back on Monday and will look into your PR first thing.

nasahlpa replied:
The test now passes on Verilator.

nasahlpa commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,243 @@
+      output_valid = false;
+      CHECK_DIF_OK(
+          dif_aes_get_status(&aes, kDifAesStatusOutputValid, &output_valid));
+      if (output_valid) {
+        LOG_INFO(
+            "AES encryption has output_valid as expected after entropy "
+            "disabled at block %d",
+            i);
+      } else {
+        LOG_ERROR("ERROR:AES encryption did not generate output_valid");
+      }
```
```suggestion
      CHECK(output_valid, "ERROR:AES encryption did not generate output_valid");
```
Please change this. With this `CHECK`, the test fails (and prints the according `FAIL` message) when `output_valid` is not true. Same for the `aes_force_prng_reseed` test.

nasahlpa commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,243 @@
+      output_valid = false;
+      CHECK_DIF_OK(
+          dif_aes_get_status(&aes, kDifAesStatusOutputValid, &output_valid));
+      if (!output_valid) {
+        LOG_INFO(
+            "AES encryption is waiting for output_valid as expected at block "
+            "%d",
+            i);
+      } else {
+        LOG_ERROR(
+            "ERROR:AES encryption generated output_valid and did not halt for "
+            "entropy to be enabled");
+      }
```
```suggestion
      CHECK(!output_valid, "ERROR:AES encryption generated output_valid and did"
                           "not halt for entropy to be enabled");
      LOG_INFO(
            "AES encryption is waiting for output_valid as expected at block "
            "%d",
            i);
```
Please change this. With this `CHECK`, the test fails (and prints the according `FAIL` message) when output_valid is not false. Same for the `aes_force_prng_reseed` test.

nasahlpa commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,243 @@
+      CHECK_STATUS_OK(disable_entropy_complex());
+      busy_spin_micros(1000);
+    }
+    // LOG_INFO("BLOCK completed: %d",i);
```
```suggestion
```

nasahlpa commented:
In order to proceed, the following things should be changed please:

- Squash the two commits
- Comply to the [rules](https://opentitan.org/book/doc/contributing/detailed_contribution_guide/index.html#what-should-commit-messages-look-like) of a commit message, especially to the line length (you could replace the links to the testplan points with `hw/top_earlgrey/data/ip/chip_aes_testplan.hjson`)
- Update the [chip_sw_aes_prng_reseed](https://opentitan.org/book/hw/top_earlgrey/data/chip_testplan.html?highlight=chip_sw_aes_interrupt_encryption#chip_sw_aes_prng_reseed) and [chip_sw_aes_force_prng_reseed](https://opentitan.org/book/hw/top_earlgrey/data/chip_testplan.html?highlight=chip_sw_aes_interrupt_encryption#chip_sw_aes_force_prng_reseed) test plan description to reflect what your test does
- Resolve the suggestions ([1](https://github.com/lowRISC/opentitan/pull/24574/files/befede4ce32d02537985ca97b40b4e7a580496b0#r1827250763), [2](https://github.com/lowRISC/opentitan/pull/24574/files/befede4ce32d02537985ca97b40b4e7a580496b0#r1827252643), [3](https://github.com/lowRISC/opentitan/pull/24574/files/befede4ce32d02537985ca97b40b4e7a580496b0#r1827252897)) and also apply them for the other test.

Afterwards I can approve the PR.

github-gcontributor commented:
> In order to proceed, the following things should be changed please:
> 
> * Squash the two commits
> * Comply to the [rules](https://opentitan.org/book/doc/contributing/detailed_contribution_guide/index.html#what-should-commit-messages-look-like) of a commit message, especially to the line length (you could replace the links to the testplan points with `hw/top_earlgrey/data/ip/chip_aes_testplan.hjson`)
> * Update the [chip_sw_aes_prng_reseed](https://opentitan.org/book/hw/top_earlgrey/data/chip_testplan.html?highlight=chip_sw_aes_interrupt_encryption#chip_sw_aes_prng_reseed) and [chip_sw_aes_force_prng_reseed](https://opentitan.org/book/hw/top_earlgrey/data/chip_testplan.html?highlight=chip_sw_aes_interrupt_encryption#chip_sw_aes_force_prng_reseed) test plan description to reflect what your test does
> * Resolve the suggestions ([1](https://github.com/lowRISC/opentitan/pull/24574/files/befede4ce32d02537985ca97b40b4e7a580496b0#r1827250763), [2](https://github.com/lowRISC/opentitan/pull/24574/files/befede4ce32d02537985ca97b40b4e7a580496b0#r1827252643), [3](https://github.com/lowRISC/opentitan/pull/24574/files/befede4ce32d02537985ca97b40b4e7a580496b0#r1827252897)) and also apply them for the other test.
> 
> Afterwards I can approve the PR.

Updated and code checked in as per your suggestion.

nasahlpa reviewed (approved):
Thanks for your changes - LGTM! Could you please resolve the linting issue detected by CI?

@engdoreis do you also want to have another look before merging this PR?

github-gcontributor commented:
> Thanks for your changes - LGTM! Could you please resolve the linting issue detected by CI?
> 
> @engdoreis do you also want to have another look before merging this PR?

Hi @nasahlpa,
Sorry.  I need some time to double check few things.  Please wait before merging this code.

github-gcontributor commented:
> > Thanks for your changes - LGTM! Could you please resolve the linting issue detected by CI?
> > 
> > @engdoreis do you also want to have another look before merging this PR?
> 
> Hi @nasahlpa,
> Sorry.  I need some time to double check few things.  Please wait before merging this code.


Hi @nasahlpa ,
I updated the aes_prng_test and .hjson to include 8kBlock test and description.
Kindly review and approve.

engdoreis commented on hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:
```diff
@@ -143,34 +143,67 @@
-            encryption stalls when reseed is required. Afterwards it should verify that AES
-            finishes encryption after EDN is re-enabled.
+      desc: ''' Verify that the AES module halts when a PRNG reseed is required
+but the entropy complex is disabled, and resumes encryption once the entropy
```
Can you please fix this alignment?

github-gcontributor replied:
Noted

github-gcontributor replied:
Noted

engdoreis commented on hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:
```diff
@@ -183,28 +216,49 @@
-      desc: '''Verify that when KEY_TOUCH_FORCES_RESEED is set, the PRNG reseed is
-               triggered every time a key changes
+      desc: '''Verify that the AES module stalls/halts when the entropy complex
+is disabled before starting encryption and enables output of encryption once
```
+ Fix alignment

github-gcontributor replied:
Noted

github-gcontributor replied:
Noted

engdoreis commented on sw/device/tests/BUILD:
```diff
@@ -151,6 +151,64 @@ opentitan_test(
+opentitan_test(
+    name = "aes_force_prng_reseed_test",
+    srcs = ["aes_force_prng_reseed_test.c"],
+    exec_env = dicts.add(
```
I think that you want to add the execution environments for sival:
https://github.com/lowRISC/opentitan/blob/c52ee256f90c67f05646e2aa3e8b0d2c51aef1a1/sw/device/tests/BUILD#L93-L100

github-gcontributor replied:
Sure, noted

engdoreis commented on sw/device/tests/BUILD:
```diff
@@ -151,6 +151,64 @@ opentitan_test(
+opentitan_test(
+    name = "aes_prng_reseed_test",
+    srcs = ["aes_prng_reseed_test.c"],
+    exec_env = dicts.add(
```
Add sival exec_envs

github-gcontributor replied:
Sure, Noted.

engdoreis commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,218 @@
+  kAesNumBlocks = 8,
+  // NumBytes in one block is 4x8 i.e. 32 bits in 1 block
+  kDifAesBlockNumBytes = 4,
+  disable_entropy_at_start_en = 1,
```
```suggestion
  kDisableEntropyAtStartEn = 1,
```

github-gcontributor replied:
Noted

engdoreis commented on sw/device/tests/aes_force_prng_reseed_test.c:
```diff
@@ -0,0 +1,218 @@
+      ": %d",
+      kAesNumBlocks, kAesNumBlocks - 1, disable_entropy_at_start_en);
+  // Initialize AES
+  CHECK_DIF_OK(
```
This function returns `status_t`, so I recommend using `TRY` rather than `CHECK_DIF_OK`

engdoreis replied:
The same applies to the other test

github-gcontributor replied:
Noted

github-gcontributor replied:
Noted

engdoreis commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,291 @@
+static dif_aes_t aes;
+dif_aes_key_share_t key;
+enum {
+  NUM_TEST_CONFIGS = 2,
```
Per coding style guide.
```suggestion
  kNumTestConfigs = 2,
```

github-gcontributor replied:
Noted

github-gcontributor replied:
Updated code and checked in at my end

engdoreis commented on sw/device/tests/aes_prng_reseed_test.c:
```diff
@@ -0,0 +1,291 @@
+  CHECK_DIF_OK(dif_rv_core_ibex_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR),
+      &rv_core_ibex));
+  for (uint32_t i = 0; i < 2; ++i) {
```
```suggestion
  for (uint32_t i = 0; i < ARRAYSIZE(plain_text); ++i) {
```

github-gcontributor replied:
Noted

github-gcontributor commented:
https://github.com/lowRISC/opentitan/pull/24574#discussion_r1832421098 
Not sure what you meant, I assume it to be consistent with the aes_prng_test and updated with random number generation logic.

github-gcontributor commented:
> > > > Thanks for your changes - LGTM! Could you please resolve the linting issue detected by CI?
> > > > @engdoreis do you also want to have another look before merging this PR?
> > > 
> > > 
> > > Hi @nasahlpa,
> > > Sorry.  I need some time to double check few things.  Please wait before merging this code.
> > 
> > 
> > Hi @nasahlpa , I updated the aes_prng_test and .hjson to include 8kBlock test and description. Kindly review and approve.
> 
> I had a look into the update - LGTM! Please address the comments from @engdoreis then we can merge it.

Suggestions are addressed and code is checked-in

engdoreis reviewed (changes requested):
@github-gcontributor thanks for the changes and fixes, it LGTM.
The second commit looks like a leftover, can you please drop it before we merge?

github-gcontributor commented:
> @github-gcontributor thanks for the changes and fixes, it LGTM. The second commit looks like a leftover, can you please drop it before we merge?

Sorry. Can you help me understand this issue more.  I see single commit at my end.  I understand it as we need to close the open issues. I updated commit message accordingly.

nasahlpa commented:
> > @github-gcontributor thanks for the changes and fixes, it LGTM. The second commit looks like a leftover, can you please drop it before we merge?
> 
> Sorry. Can you help me understand this issue more. I see single commit at my end. I understand it as we need to close the open issues. I updated commit message accordingly.

If you go to the "Commits" tab here on GitHub you'll find two commits:
![Screenshot_20241114-200935](https://github.com/user-attachments/assets/39d1439c-40a9-4676-9b89-10dc62506bbc)

github-gcontributor commented:
> > > @github-gcontributor thanks for the changes and fixes, it LGTM. The second commit looks like a leftover, can you please drop it before we merge?
> > 
> > 
> > Sorry. Can you help me understand this issue more. I see single commit at my end. I understand it as we need to close the open issues. I updated commit message accordingly.
> 
> If you go to the "Commits" tab here on GitHub you'll find two commits: ![Screenshot_20241114-200935](https://private-user-images.githubusercontent.com/5767400/386336731-39d1439c-40a9-4676-9b89-10dc62506bbc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE2MzMyMzEsIm5iZiI6MTczMTYzMjkzMSwicGF0aCI6Ii81NzY3NDAwLzM4NjMzNjczMS0zOWQxNDM5Yy00MGE5LTQ2NzYtOWI4OS0xMGRjNjI1MDZiYmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTVUMDEwODUxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjAzMDE3YmEyNDliYWQ0ZDFhZDRlMzI1MWJhNzViZjBkZmVmYzI5MmEwZTBlNTliNGNhOWVjOWQ5OTllZGQwZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ilQuq3Gr7w4arbJj8u_D8TdAv-Lq2xFDaUBTolplFBM)

Thanks for pointing it out.
git rebase HEAD ~2 did not work for me (shows 923 commits to squash and pick)

Tried below:
git log --oneline -n 3
d65a51afe6 (HEAD -> vt_checkins, origin/vt_checkins) Merge branch 'master' into vt_checkins
f191f1085c (temp_branch) [sival/aes] Add aes_prng_reseed,aes_prng_force_reseed
9da36ca26c [bazel] add rule to assemble several binaries

git branch vt_checkins_backup (created backup)
git reset --hard f191f1085c (reset to f191f1085c )
git log --oneline -n 3
f191f1085c (HEAD -> vt_checkins) [sival/aes] Add aes_prng_reseed,aes_prng_force_reseed
0df3194acc (origin/master, origin/HEAD) [sram_ctrl,dv] Add readback_regwen test
db48ef65ff [top_englishbreakfast] Add second UART

git push -f origin vt_checkins
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
To github.com:github-gcontributor/opentitan.git
 + d65a51afe6...f191f1085c vt_checkins -> vt_checkins (forced update)

This should resolve the two commits.  Let me know your inputs on this.

nasahlpa commented:
Hm here on Github I still see two commits.

engdoreis commented:
This is what you can do to remove the second commit:
```sh
git fetch --all --prune
git checkout vt_checkins
git reset --hard origin/vt_checkins
# This should remove the most recent commit. Make sure that the most recent is the one that you want deleted.
git reset --hard HEAD^
```
Inspect the branch and if you're happy you can for push to this PR
```sh
git push origin vt_checkings -f
```

github-gcontributor commented:
> ```shell
> git fetch --all --prune
> git checkout vt_checkins
> git reset --hard origin/vt_checkins
> ```

Thanks.
While my attempt (https://github.com/lowRISC/opentitan/pull/24574#issuecomment-2477762580) using linux environment did not work,  I tried below on Codespace "https://urban-acorn-v6gp6rvjqp5qhpvg4.github.dev/":
git log --oneline -n 3
    d65a51afe6 (HEAD -> vt_checkins) Merge branch 'master' into vt_checkins
    **f191f1085c** (origin/vt_checkins) [sival/aes] Add aes_prng_reseed,aes_prng_force_reseed
    9da36ca26c [bazel] add rule to assemble several binaries

git fetch --all --prune
git checkout vt_checkins
git reset --hard origin/vt_checkins
git reset --hard HEAD^ --> got me one level down
git log --oneline --all | grep "f191f1085c" --> could find my commit [aes/sival] ... 
git reset --hard f191f1085c  --> Reached to my commit
git push -f origin vt_checkins 

Output of the last command:
Everything up-to-date
And the second commit is still present!  Is it a stale copy?

engdoreis commented:
@github-gcontributor could you please handle this cherry-pick?

github-gcontributor commented:
HI @engdoreis,
the backport is complete:

git push origin backport-24574-to-earlgrey_1.0.0
Enumerating objects: 23, done.
Counting objects: 100% (23/23), done.
Delta compression using up to 12 threads
Compressing objects: 100% (13/13), done.
Writing objects: 100% (13/13), 6.37 KiB | 724.00 KiB/s, done.
Total 13 (delta 11), reused 0 (delta 0), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (11/11), completed with 10 local objects.
remote: 
remote: Create a pull request for 'backport-24574-to-earlgrey_1.0.0' on GitHub by visiting:
remote:      https://github.com/github-gcontributor/opentitan/pull/new/backport-24574-to-earlgrey_1.0.0
remote: 
To github.com:github-gcontributor/opentitan.git
 * [new branch]            backport-24574-to-earlgrey_1.0.0 -> backport-24574-to-earlgrey_1.0.0

Kindly verify and proceed for the next step or let me know if anything from my end.
Thanks.

engdoreis commented:
@github-gcontributor can you please create and link the backport PR here?

github-gcontributor commented:
> @github-gcontributor can you please create and link the backport PR here?

Crearted backport-24574-to-earlgrey_1.0.0 
Please suggest the next steps

engdoreis commented:
@github-gcontributor I can see that you created a branch on your fork named `backport-24574-to-earlgrey_1.0.0`. Now you have to create a pull request from that branch to `earlgrey_1.0.0` and add the link of this PR on the description.

Please see this example https://github.com/lowRISC/opentitan/pull/25399

github-gcontributor commented:
> @github-gcontributor I can see that you created a branch on your fork named `backport-24574-to-earlgrey_1.0.0`. Now you have to create a pull request from that branch to `earlgrey_1.0.0` and add the link of this PR on the description.
> 
> Please see this example #25399

https://github.com/lowRISC/opentitan/pull/25441
